mesh/mesh.go

```diff
@@ -431,133 +448,128 @@ func (msh *Mesh) revertState(ctx context.Context, layerID types.LayerID) error {
-		log.Int("valid_block_txs", len(validBlockTxs)),
-		l.Index(),
+		block.LayerIndex,
+		log.Int("valid_block_txs", len(txs)),
 		log.Int("num_failed_txs", len(failedTxs)),
 	)
-
-	msh.setLatestLayerInState(l.Index())
+	return nil
 }
 
 // ProcessLayerPerHareOutput receives hare output once it finishes running for a given layer.
-func (msh *Mesh) ProcessLayerPerHareOutput(ctx context.Context, validatedLayer types.LayerID, blockIDs []types.BlockID) {
-	logger := msh.WithContext(ctx).WithFields(validatedLayer)
-
-	// when ProcessLayerPerHareOutput is called by hare, we may not have all the blocks in the hare output.
-	// so try harder to fetch the blocks from peers.
-	if len(blockIDs) > 0 && !validatedLayer.GetEpoch().IsGenesis() {
-		if err := msh.fetch.GetBlocks(ctx, blockIDs); err != nil {
-			logger.With().Warning("not all blocks are available locally", validatedLayer, log.Err(err))
-		}
-	}
-
-	var blocks []*types.Block
-	for _, blockID := range blockIDs {
-		block, err := msh.GetBlock(blockID)
+func (msh *Mesh) ProcessLayerPerHareOutput(ctx context.Context, layerID types.LayerID, blockID types.BlockID) error {
+	logger := msh.WithContext(ctx).WithFields(layerID, blockID)
+	if blockID == types.EmptyBlockID {
+		logger.Info("received empty set from hare")
+	} else {
+		// double-check we have this block in the mesh
+		_, err := msh.GetBlock(blockID)
 		if err != nil {
-			// stop processing this hare result, wait until tortoise pushes this layer into state
-			logger.With().Error("hare terminated with block that is not present in mesh",
-				blockID,
-				log.Err(err))
-			return
+			logger.With().Error("hare terminated with block that is not present in mesh", log.Err(err))
+			return err
 		}
-		blocks = append(blocks, block)
 	}
-
-	metrics.LayerNumBlocks.WithLabelValues().Observe(float64(len(blockIDs)))
-
 	// report that hare "approved" this layer
 	events.ReportLayerUpdate(events.LayerUpdate{
-		LayerID: validatedLayer,
+		LayerID: layerID,
 		Status:  events.LayerStatusTypeApproved,
 	})
 
-	logger.With().Info("saving hare output for layer", log.Int("num_hare_output", len(blocks)))
-	if err := msh.SaveHareConsensusOutput(ctx, validatedLayer, types.BlockIDs(blocks)); err != nil {
-		logger.With().Error("saving hare output failed", log.Err(err))
+	logger.Info("saving hare output for layer")
+	if err := msh.SaveHareConsensusOutput(ctx, layerID, blockID); err != nil {
+		logger.Error("saving layer hare output failed")
 	}
-	lyr := types.NewExistingLayer(validatedLayer, blocks)
-	msh.ProcessLayer(ctx, lyr)
+	return msh.ProcessLayer(ctx, layerID)
 }
 
-// apply the state for a single layer. stores layer results for early layers, and applies previously stored results for
-// now-older layers.
-func (msh *Mesh) updateStateWithLayer(ctx context.Context, layer *types.Layer) {
+// apply the state for a single layer.
+func (msh *Mesh) updateStateWithLayer(ctx context.Context, layerID types.LayerID, hareOutput *types.Block) error {
```

Comment:
i think it doesn't make sense to call it hare output after tortoise verified a block

Reply:
changed to `block`